Exchange/handler.ts

```diff
@@ -0,0 +1,81 @@
+import * as express from "express";
+import * as E from "fp-ts/Either";
+import * as TE from "fp-ts/TaskEither";
+import { pipe } from "fp-ts/lib/function";
+
+import { ContextMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/context_middleware";
+import {
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "@pagopa/io-functions-commons/dist/src/utils/request_middleware";
+import { readableReportSimplified } from "@pagopa/ts-commons/lib/reporters";
+import {
+  IResponseErrorInternal,
+  IResponseSuccessJson,
+  ResponseErrorInternal,
+  ResponseSuccessJson
+} from "@pagopa/ts-commons/lib/responses";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+import { defaultLog } from "@pagopa/winston-ts";
+
+import { IConfig } from "../utils/config";
+
+import { ExchangeToken } from "../generated/definitions/external/ExchangeToken";
+import {
+  MagicLinkPayload,
+  getGenerateExchangeJWT
+} from "../utils/exchange-jwt";
+import { BaseJwePayload } from "../utils/jwe";
+import { magicLinkJweValidationMiddleware } from "../utils/middlewares/magic-link-jwe-validation-middleware";
+
+type ExchangeHandlerT = (
+  user: BaseJwePayload
+) => Promise<IResponseErrorInternal | IResponseSuccessJson<ExchangeToken>>;
+
+const EXCHANGE_TOKEN_TYPE = "exchange" as NonEmptyString;
+
+export const exchangeHandler = (config: IConfig): ExchangeHandlerT => (
+  user_data: BaseJwePayload
+): ReturnType<ExchangeHandlerT> =>
+  pipe(
+    user_data,
+    MagicLinkPayload.decode,
+    E.mapLeft(e =>
+      ResponseErrorInternal(
+        `Invalid Magic Link JWE: ${readableReportSimplified(e)}`
+      )
+    ),
+    TE.fromEither,
+    TE.chain(magic_link_payload =>
+      pipe(
+        getGenerateExchangeJWT(config)({
+          ...magic_link_payload,
+          token_type: EXCHANGE_TOKEN_TYPE
+        }),
+        TE.mapLeft(e =>
+          ResponseErrorInternal(
+            `Cannot generate Exchange JWT|ERROR=${e.message}`
+          )
+        ),
+        defaultLog.taskEither.errorLeft(
+          r => r.detail ?? "Cannot generate Exchange JWT"
+        ),
+        TE.map(jwt => ResponseSuccessJson({ jwt } as ExchangeToken))
+      )
+    ),
+    TE.toUnion
+  )();
```

Comment:
```suggestion
export const exchangeHandler = (config: IConfig): ExchangeHandlerT => (
  user_data: BaseJwePayload
): ReturnType<ExchangeHandlerT> =>
  pipe(
    getGenerateExchangeJWT(config)({
      family_name: user_data.family_name,
      fiscal_number: user_data.fiscal_number,
      name: user_data.name,
      token_type: TokenTypes.EXCHANGE
    }),
    TE.mapLeft(e =>
      ResponseErrorInternal(`Cannot generate Exchange JWT|ERROR=${e.message}`)
    ),
    defaultLog.taskEither.errorLeft(
      r => r.detail ?? "Cannot generate Exchange JWT"
    ),
    TE.map(ResponseSuccessJson),
    TE.toUnion
  )();
```
I would simplify this handler as I suggested above because there are some things that aren't needed. let me know what you think

Reply:
Agree with @arcogabbo